hail/src/main/scala/is/hail/expr/ir/Parser.scala

```diff
@@ -1450,13 +1363,29 @@ object IRParser {
           rngState <- ir_value_expr(env)(it)
           args <- ir_value_children(env)(it)
         } yield ApplySeeded(function, args, rngState, staticUID, rt)
-      case "ApplyIR" | "ApplySpecial" | "Apply" =>
+      case "ApplyIR" =>
         val errorID = int32_literal(it)
         val function = identifier(it)
         val typeArgs = type_exprs(it)
         val rt = type_expr(it)
         ir_value_children(env)(it).map { args =>
-          invoke(function, rt, typeArgs, errorID, args: _*)
+          ApplyIR(function, rt, typeArgs, args, errorID)
+        }
+      case "ApplySpecial" =>
+        val errorID = int32_literal(it)
+        val function = identifier(it)
+        val typeArgs = type_exprs(it)
+        val rt = type_expr(it)
+        ir_value_children(env)(it).map { args =>
+          ApplySpecial(function, typeArgs, args, rt, errorID)
+        }
+      case "Apply" =>
+        val errorID = int32_literal(it)
+        val function = identifier(it)
+        val typeArgs = type_exprs(it)
+        val rt = type_expr(it)
+        ir_value_children(env)(it).map { args =>
+          Apply(function, typeArgs, args, rt, errorID)
         }
```

Comment:
This is a little sad. Can you define a function that takes the data constructor to clean all this up? ie
```scala
def applyLikeParser[Ap <: IR](cons: (String, Type, IndexedSeq[Type], IndexedSeq[IR], Int) => Ap): Parser[Ap] =
  /* The de-duplicated implementation */

val parseApply: Parser[Apply] = 
  applyLikeParser(Apply)
val parseApplySpecial: Parser[ApplySpecial] = 
  applyLikeParser(ApplySpecial)
...

case "Apply" =>
  parseApply(env)(it)
case "ApplySpecial" =>
  parseApplySpecial(env)(it)
...
```